libp2p/services/autonatservice.nim

```diff
@@ -0,0 +1,128 @@
+# Nim-LibP2P
+# Copyright (c) 2022 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import std/[options, tables]
+import chronos
+import ../switch
+import ../protocols/[connectivity/autonat,
+                    rendezvous]
+import ../protocols/connectivity/relay/[relay, client]
+import ../discovery/[rendezvousinterface, discoverymngr]
+import ../utils/heartbeat
+import ../crypto/crypto
+
+type
+  AutonatService* = ref object of Service
+    registerLoop: Future[void]
+    scheduleInterval: Option[Duration]
+    networkReachability: NetworkReachability
+    t: CountTable[NetworkReachability]
```

Comment:
I believe this state must outlive `run` calls. Am I missing something?

Reply:
Ah perhaps im missing something. What I had in mind is every `interval` we ask `x` peers and if `y` flags us as `Reachable`, that's it. So in the next interval, I think we don't need the `t`.

Of course I'm assuming that we do something like [this](https://github.com/status-im/nim-libp2p/pull/814#discussion_r1042073528). So within the same function we await until all peers have replied.

Reply:
Now I understand what you mean. If in a run we ask 5 peers, 2 say NotReachable, 2 say Reachable and 1 says Unknown. Should we keep the current status, discard all answers and start from scratch in the next run? 

Btw I think my current algo is wrong. It would end up with an empty table in the case above and it's not what we probably want.